src/Language/PureScript/CoreFn/Laziness.hs

```diff
@@ -0,0 +1,504 @@
+-- here with a Maybe. In a function application `f a b`, `f` has force 2, but
+-- `a` and `b` have unknown force--it depends on what `f` does with them.
+--
+-- The rules of assigning delay and force are simple:
+--   * The expressions that are assigned to bindings in this group have
+--     delay 0, force 0.
+--   * In a function application, the function expression has force 1 higher
+--     than the force of the application expression, and the argument
+--     expression has unknown force.
+--     * UNLESS this argument is being directly provided to a constructor (in
+--       other words, the function expression is either a constructor itself or
+--       a constructor that has already been partially applied), in which case
+--       the force of both subexpressions is unchanged. We can assume that
+--       constructors don't apply any additional force to their arguments.
+--   * If the force of a lambda is zero, the delay of the body of the lambda is
+--     incremented; otherwise, the force of the body of the lambda is
+--     decremented. (Applying one argument to a lambda cancels out one unit of
+--     delay.)
+--   * In the argument of a Case and the bindings of a Let, force is unknown.
+--   * Everywhere else, preserve the delay and force of the enclosing
+--     expression.
+--
+-- (Note that this analysis is quite ignorant of any actual control flow
+-- choices made at run time. It doesn't even track what happens to a reference
+-- after it has been locally bound by a Let or Case. Instead, it just assumes
+-- the worst--once locally bound to a new name, it imagines that absolutely
+-- anything could happen to that new name and thus to the underlying reference.
+-- But the value-to-weight ratio of this approach is perhaps surprisingly
+-- high.)
+--
+-- Every subexpression gets a delay and a force, but we are only interested
+-- in references to other bindings in the binding group, so the traversal only
+-- exposes `Var`s to the provided function.
+--
+onVarsWithDelayAndForce :: forall f. Applicative f => (Int -> Maybe Int -> Ann -> Qualified Ident -> f (Expr Ann)) -> Expr Ann -> f (Expr Ann)
+onVarsWithDelayAndForce f = snd . go 0 $ Just 0
+  where
+  go :: Int -> Maybe Int -> (Bind Ann -> f (Bind Ann), Expr Ann -> f (Expr Ann))
+  go delay force = (handleBind, handleExpr')
+    where
+    (handleBind, handleExpr, handleBinder, handleCaseAlternative) = traverseCoreFn handleBind handleExpr' handleBinder handleCaseAlternative
+    handleExpr' = \case
+      Var a i -> f delay force a i
+      Abs a i e -> Abs a i <$> snd (if force == Just 0 then go (succ delay) force else go delay $ fmap pred force) e
+      -- A clumsy hack to preserve TCO in a particular idiom of unsafePartial seen in Data.Map.Internal.
+      App a1 e1@(Var _ (Qualified (Just C.PartialUnsafe) (Ident up))) (Abs a2 i e2) | up == C.unsafePartial
+        -> App a1 e1 . Abs a2 i <$> handleExpr' e2
```

Comment:
I think that, generally, the expectation around optimizations is that inlining opens up more opportunities for it. The advantage is that with a decent inliner, more things can be considered tail recursive. The disadvantage is that relying on optimizations for something like stack safety can be quite brittle.

Take for example, existentials. We can CPS encode existentials (or use `purescript-exists`), but that means it's _impossible_ to write safe, fast, tail-recursive code with them. I've run up against this so, so many times and it's incredibly frustrating. Language support would be nice, but another alternative is if the eliminator were inlined (since it's the equivalent of aliasing `unsafeCoerce`), it would just work. Another option is `tailRec`, but it has significant overhead. If we could totally eliminate the overhead of `tailRec`, I might be OK with removing TCO altogether.

That doesn't really answer the question, except to say:
* Given the status quo, I think we should grant the expectation that unsafeCoerce/unsafePartial are removed.
* If we only had optimized `tailRec` (with multiple parameters), the issue would be moot because then it's no longer a correctness problem, but truly an optimization problem.

Reply:
I like the sound of that, but to be clear, when talking about optimizing `tailRec`, do you mean the case of `tailRec (something that inlines to a literal lambda where there is a `Loop` or `Done` lexically present in every tail position) initValue`, or the case of `tailRec (arbitrary expression) initValue`? Because the first sounds straightforward and lovely, and the second sounds impossible unless you have other special cases in mind that you would carve out.

Assuming we agree on what the above means, the logical plan would then be keep the hacks in this analysis for 0.15, implement `tailRec` optimization for 0.16 and warn everyone to use it if they care about stack safety, remove the hacks in 0.17?

Reply:
Yes, I mean the former. `tailRec` is nice because if the optimization bails it's still stack safe, but it may allocate more, so it's not an issue of something working vs not working, just how efficient it is. I don't know if we can realistically get rid of TCO in general (tailRec is implemented in terms of TCO, after all), but we can dissuade it's usage, and basically say that it's necessary if you want to guarantee stack safety.